### DISCUSSION TITLE
NRT - Deeper analysis inside constructor for indirect intialization scenarios

### DISCUSSION THREAD

**Discussion by artdedeco** [dotnet/csharplang#2869] [Language Ideas]
The `InitializeComponent` scenario:

```cs
public class Foo {
    public Foo() { // warning here about _x not being initialized
        InitializeComponent();
    }

    public InitializeComponent() {
         _x = "Foo";
         // initialize fields and stuff
    }

    private string _x;
}
```

Also, happened to me on the first project I'm updating with NRTs:

```cs
public class Foo {
    public Foo(Uri baseUrl) {
        baseUrl = null!; // current workaround
        BaseUrl = baseUrl;
    }

    private Uri _baseUrl;
    public Uri BaseUrl { get => _baseUrl; set { /* some logic*/ _baseUrl = value; } }
}
```

The compiler should follow on the method calls inside a constructor to a reasonable depth in order to determine nullability.

**Comment by Joe4evr** [comment]
Already pointed this out on gitter, but related to #2328.

**Comment by jnm2** [comment]
This is coming up frequently enough to be annoying across every codebase (open-source and proprietary) that I've used NRTs with so far. I've been doing `= null!`.

Flow control could in theory prove that the constructor was calling a method or property accessor which unconditionally initializes that field.
